Modules/_posixsubprocess.c

```diff
@@ -561,26 +589,33 @@ subprocess_fork_exec(PyObject* self, PyObject *args)
     PyObject *env_list, *preexec_fn;
     PyObject *process_args, *converted_args = NULL, *fast_args = NULL;
     PyObject *preexec_fn_args_tuple = NULL;
+    PyObject *groups_list;
     int p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite;
     int errpipe_read, errpipe_write, close_fds, restore_signals;
     int call_setsid;
+    int call_setgid, call_setgroups, call_setuid;
+    uid_t uid;
+    gid_t gid, *groups = NULL;
     PyObject *cwd_obj, *cwd_obj2;
     const char *cwd;
     pid_t pid;
     int need_to_reenable_gc = 0;
     char *const *exec_array, *const *argv = NULL, *const *envp = NULL;
-    Py_ssize_t arg_num;
+    Py_ssize_t arg_num, num_groups = 0;
     int need_after_fork = 0;
     int saved_errno = 0;
 
     if (!PyArg_ParseTuple(
-            args, "OOpO!OOiiiiiiiiiiO:fork_exec",
+            args, "OOpO!OOiiiiiiiiiiiiiOiiO:fork_exec",
             &process_args, &executable_list,
             &close_fds, &PyTuple_Type, &py_fds_to_keep,
             &cwd_obj, &env_list,
             &p2cread, &p2cwrite, &c2pread, &c2pwrite,
             &errread, &errwrite, &errpipe_read, &errpipe_write,
-            &restore_signals, &call_setsid, &preexec_fn))
+            &restore_signals, &call_setsid,
+            &call_setgid, &gid, &call_setgroups, &groups_list,
+            &call_setuid, &uid,
```

Comment:
Here we have the same problem as with `INT_MAX` bound on `extra_groups` below. If there is a consensus that we can use `_Py_Uid_Converter/_Py_Gid_Converter`, may be we could receive an object and convert it afterwards?

Note that we potentially also have an opposite problem here: if `sizeof(uid_t) < sizeof(int)`, we'll overflow stack-allocated variables. I don't know a modern system where that would be true, but Linux had 16-bit uids before 2.4.

Reply:
I migrated to passing in as an object and using _Py_Uid_Converter/_Py_Gid_Converter